// geotiff/dist-node/logging.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setLogger = setLogger;
exports.debug = debug;
exports.log = log;
exports.info = info;
exports.warn = warn;
exports.error = error;
exports.time = time;
exports.timeEnd = timeEnd;
/**
 * A no-op logger
 */
class DummyLogger {
    log(..._args) { }
    debug(..._args) { }
    info(..._args) { }
    warn(..._args) { }
    error(..._args) { }
    time(..._args) { }
    timeEnd(..._args) { }
}
let LOGGER = new DummyLogger();
/**
 *
 * @param {object} logger the new logger. e.g `console`
 */
function setLogger(logger = new DummyLogger()) {
    LOGGER = logger;
}
function debug(...args) {
    return LOGGER.debug(...args);
}
function log(...args) {
    return LOGGER.log(...args);
}
function info(...args) {
    return LOGGER.info(...args);
}
function warn(...args) {
    return LOGGER.warn(...args);
}
function error(...args) {
    return LOGGER.error(...args);
}
function time(...args) {
    return LOGGER.time(...args);
}
function timeEnd(...args) {
    return LOGGER.timeEnd(...args);
}
