// eleventy-plugin-vento/dist/plugin.mjs
import path from "node:path";
import autotrimPlugin, { defaultTags } from "ventojs/plugins/auto_trim.js";
import createVentoEnv from "ventojs";
import createDebugger from "debug";

//#region src/utils/create-vento-tag.ts
function createVentoTag(tagInfo) {
	let LEVEL = 1;
	const IS_PAIRED = tagInfo.group === "pairedShortcodes";
	const tag = (env, token, output, tokens) => {
		const code = token[1];
		if (!(code === tagInfo.name || code.startsWith(`${tagInfo.name} `))) return;
		const { dataVarname } = env.options;
		const thisArg = `{
			page: ${dataVarname}.page,
			eleventy: ${dataVarname}.eleventy
		}`;
		const fn = `__env.utils.eleventyFunctions.${tagInfo.group}.${tagInfo.name}`;
		const args = [code.replace(tagInfo.name, "").trim()];
		const compiled = [];
		if (IS_PAIRED) {
			const nestedVarname = `__content${output.startsWith("__content") ? LEVEL++ : 0}`;
			args.unshift(env.compileFilters(tokens, nestedVarname, env.options.autoescape));
			compiled.push("{", `let ${nestedVarname} = "";`, ...env.compileTokens(tokens, nestedVarname, `/${tagInfo.name}`));
		}
		args.unshift(thisArg);
		compiled.push("{", `const __result = await ${fn}.call(${args.filter(Boolean).join(", ")});`, `${output} += ${env.compileFilters(tokens, "__result", env.options.autoescape)}`, "}");
		if (IS_PAIRED) compiled.push("}");
		return compiled.join("\n");
	};
	return Object.defineProperty(tag, "name", { value: `${tagInfo.name}_${IS_PAIRED ? `EleventyPairedTag` : `EleventyTag`}` });
}

//#endregion
//#region src/utils/debuggers.ts
const debugBaseNamespace = "Eleventy:Vento";
const debugMain = createDebugger(debugBaseNamespace);
const debugCache = createDebugger(`${debugBaseNamespace}:Cache`);
const debugRender = createDebugger(`${debugBaseNamespace}:Render`);

//#endregion
//#region src/engine.ts
function createVentoEngine(options) {
	const env = createVentoEnv(options);
	env.utils.eleventyFunctions = {
		shortcodes: {},
		pairedShortcodes: {}
	};
	function loadPlugins(plugins) {
		for (const plugin of plugins) env.use(plugin);
	}
	function loadFilters(filters) {
		for (const [name, fn] of Object.entries(filters)) env.filters[name] = function(...args) {
			const { page, eleventy } = this.data;
			return fn.apply({
				...this,
				page,
				eleventy
			}, args);
		};
	}
	function loadShortcodes(shortcodes) {
		for (const [name, fn] of Object.entries(shortcodes)) {
			env.utils.eleventyFunctions.shortcodes[name] = fn;
			const shortcodeTag = createVentoTag({
				name,
				group: "shortcodes"
			});
			env.tags.push(shortcodeTag);
		}
	}
	function loadPairedShortcodes(pairedShortcodes) {
		for (const [name, fn] of Object.entries(pairedShortcodes)) {
			env.utils.eleventyFunctions.pairedShortcodes[name] = fn;
			const shortcodeTag = createVentoTag({
				name,
				group: "pairedShortcodes"
			});
			env.tags.push(shortcodeTag);
		}
	}
	async function getTemplateFunction(source, file, useVentoCache = true) {
		let template = await env.cache.get(file);
		if (template?.source === source) debugCache("Cache HIT for `%s`, used precompiled template", file);
		else {
			template = env.compile(source, file);
			debugCache(`Cache MISS for \`%s\`, compiled new template; %s\n\n${template}`, file, useVentoCache ? "let Vento cache result" : "defer result caching to Eleventy");
			if (useVentoCache) env.cache.set(file, template);
		}
		return template;
	}
	return {
		cache: env.cache,
		loadPlugins,
		loadFilters,
		loadShortcodes,
		loadPairedShortcodes,
		getTemplateFunction
	};
}
async function renderVentoTemplate(template, data, from) {
	debugRender("Rendering `%s`", from);
	const { content } = await template(data);
	return content;
}

//#endregion
//#region src/plugin.ts
function VentoPlugin(eleventyConfig, userOptions) {
	debugMain("Initializing eleventy-plugin-vento");
	eleventyConfig.versionCheck(">=3.0.0");
	const options = {
		autotrim: false,
		plugins: [],
		filters: true,
		shortcodes: true,
		pairedShortcodes: true,
		ventoOptions: { includes: eleventyConfig.directories.includes },
		...userOptions
	};
	debugMain("Merged default and user config: %O", options);
	const filters = eleventyConfig.getFilters();
	debugMain("Reading filters from Eleventy: %o", filters);
	const shortcodes = eleventyConfig.getShortcodes();
	debugMain("Reading shortcodes from Eleventy: %o", shortcodes);
	const pairedShortcodes = eleventyConfig.getPairedShortcodes();
	debugMain("Reading paired shortcodes from Eleventy: %o", pairedShortcodes);
	if (options.autotrim) {
		const tagSet = new Set(options.autotrim === true ? ["@vento", "@11ty"] : options.autotrim);
		if (tagSet.has("@vento")) {
			tagSet.delete("@vento");
			for (const name of defaultTags) tagSet.add(name);
		}
		if (tagSet.has("@11ty")) {
			tagSet.delete("@11ty");
			for (const name in pairedShortcodes) tagSet.add(name);
		}
		debugMain("Enabled autotrim for tags: %o", tagSet);
		options.plugins.push(autotrimPlugin({ tags: [...tagSet] }));
	}
	debugMain("Initializing Vento environment");
	const engine = createVentoEngine(options.ventoOptions);
	debugMain("Loading plugins: %o", options.plugins);
	engine.loadPlugins(options.plugins);
	if (options.filters) {
		debugMain("Loading filters: %o", filters);
		engine.loadFilters(filters);
	}
	if (options.shortcodes) {
		debugMain("Loading shortcodes: %o", shortcodes);
		engine.loadShortcodes(shortcodes);
	}
	if (options.pairedShortcodes) {
		debugMain("Loading paired shortcodes: %o", pairedShortcodes);
		engine.loadPairedShortcodes(pairedShortcodes);
	}
	debugMain("Registering Vento cache handler on eleventy.beforeWatch event");
	eleventyConfig.on("eleventy.beforeWatch", async (updatedFiles) => {
		for (let file of updatedFiles) {
			file = path.normalize(file);
			debugCache("Delete cache entry for %s", file);
			engine.cache.delete(file);
		}
	});
	debugMain("Registering .vto as a template format");
	eleventyConfig.addTemplateFormats("vto");
	debugMain("Registering .vto extension with eleventy");
	eleventyConfig.addExtension("vto", {
		outputFileExtension: "html",
		read: true,
		async compile(inputContent, inputPath) {
			inputPath = path.normalize(inputPath);
			debugMain("Getting template function for `%s`", inputPath);
			const template = await engine.getTemplateFunction(inputContent, inputPath, false);
			return async (data) => await renderVentoTemplate(template, data, inputPath);
		},
		compileOptions: { async permalink(permalinkContent, inputPath) {
			if (typeof permalinkContent !== "string" || !/\{\{\s+.+\s+\}\}/.test(permalinkContent)) return permalinkContent;
			inputPath = "EleventyVentoDynamicPermalink:".concat(path.normalize(inputPath));
			debugMain("Getting template function for `%s`", inputPath);
			const template = await engine.getTemplateFunction(permalinkContent, inputPath);
			return async (data) => await renderVentoTemplate(template, data, inputPath);
		} }
	});
	debugMain("eleventy-plugin-vento initialized");
}

//#endregion
export { VentoPlugin };
